tests/unit/lms/services/lis_result_sourcedid_test.py

```diff
@@ -1,187 +1,191 @@
 
-    db_session.add(lis_result_sourcedid_)
-    return lis_result_sourcedid_
+        svc.upsert_from_request(pyramid_request, h_user, lti_user)
+
+        assert self.get_inserted_row(db_session) is None
+
+    @pytest.mark.parametrize("param", ("tool_consumer_info_product_family_code",))
+    def test_it_works_fine_with_optional_parameter_missing(
+        self, svc, pyramid_request, h_user, lti_user, db_session, param
+    ):
+        del pyramid_request.POST[param]
+
+        svc.upsert_from_request(pyramid_request, h_user, lti_user)
+
+        assert self.get_inserted_row(db_session) == Any.instance_of(LISResultSourcedId)
+
+    @classmethod
+    def model_as_dict(cls, model):
+        return {col.key: getattr(model, col.key) for col in model.iter_columns()}
+
+    @classmethod
+    def get_inserted_row(cls, db_session):
+        return db_session.query(LISResultSourcedId).one_or_none()
+
+    @pytest.fixture
+    def pyramid_request(self, pyramid_request, lti_params):
+        pyramid_request.POST.update(lti_params)
+
+        return pyramid_request
 
 
 @pytest.fixture
 def lti_user():
-    return values.LTIUser("TEST_USER_ID", "TEST_OAUTH_CONSUMER_KEY", "TEST_ROLES")
+    return LTIUser("TEST_USER_ID", "TEST_OAUTH_CONSUMER_KEY", "TEST_ROLES")
 
 
 @pytest.fixture
 def h_user():
-    return values.HUser(
+    return HUser(
         authority="TEST_AUTHORITY", username="seanh", display_name="Sample Student"
     )
 
 
 @pytest.fixture
-def application_instance(db_session):
-    """The ApplicationInstance that the LISResultSourcedIds belong to"""
-    application_instance = ApplicationInstance(
-        consumer_key="test_consumer_key",
-        shared_secret="test_shared_secret",
-        lms_url="test_lms_url",
-        requesters_email="test_requesters_email",
-    )
-    db_session.add(application_instance)
-    return application_instance
-
-
-@pytest.fixture
-def context(h_user):
+def svc(pyramid_request):
     context = mock.create_autospec(
         LTILaunchResource, spec_set=True, instance=True, h_user=h_user
     )
-    return context
-
 
-@pytest.fixture
-def svc(context, pyramid_request):
     return LISResultSourcedIdService(context, pyramid_request)
 
 
 @pytest.fixture
-def lis_result_sourcedid_info(application_instance):
-    return values.LISResultSourcedId(
-        lis_result_sourcedid="result_sourcedid",
-        lis_outcome_service_url="https://somewhere.else",
-        context_id="random context",
-        resource_link_id="random resource link id",
-        tool_consumer_info_product_family_code="MyFakeLTITool",
-    )
+def lti_params():
+    return {
+        "lis_result_sourcedid": "result_sourcedid",
+        "lis_outcome_service_url": "https://somewhere.else",
+        "context_id": "random context",
+        "resource_link_id": "random resource link id",
+        "tool_consumer_info_product_family_code": "MyFakeLTITool",
+    }
+
+
+def _add_users(lis_result_sourcedid, h_user, lti_user):
```

Comment:
I've made the change, but I think I'd like to understand why we have a different standard for our tests as our main code. If this code was used the same way in the service we would not mark it as public.

Reply:
I think it comes down to whether or not the code in question has a public API. I don't think tests really have a public API. No code that anyone writes imports and uses things from the tests. pytest just automatically calls all the `Test*` classes, `test_*()` methods, `@pytest.fixtures`, etc. So there doesn't seem any reason to indicate private vs public.

I think the same probably goes for the main code as well, in places where the main code doesn't have a public API. For example an `<APP>/views/` package just registers views for Pyramid to call, doesn't provide any names for any code that anyone writes to import. It might also apply to some other packages.

But it is important to understand that when a main code package _does_ have public names for other parts of our code to import, then it does need to indicate public vs private, and that probably applies to most of the main code.